From 949edf5582b7b0a3cb53f86f6055b263396aa7f7 Mon Sep 17 00:00:00 2001
From: Todd Whiteman <todd.whiteman@joyent.com>
Date: Tue, 5 Sep 2017 09:42:26 -0700
Subject: [PATCH] joyent/sdc-imgapi#13 imgapi crash when pulling from docker.io
 with invalid credentials Reviewed by: Trent Mick <trentm@gmail.com> Approved
 by: Trent Mick <trentm@gmail.com>

---
 CHANGES.md    |  4 ++++
 lib/errors.js | 20 +++++++++++++-------
 package.json  |  2 +-
 3 files changed, 18 insertions(+), 8 deletions(-)

diff --git a/CHANGES.md b/CHANGES.md
index 2b0e9fd..49d1fd5 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,5 +1,9 @@
 # IMGAPI changelog
 
+## 4.0.4
+
+- joyent/sdc-imgapi#13 imgapi crash when pulling from docker.io with invalid credentials
+
 ## 4.0.3
 
 - IMGAPI-637: correctly report when docker image is update to date
diff --git a/lib/errors.js b/lib/errors.js
index 96cd376..f5b45f9 100644
--- a/lib/errors.js
+++ b/lib/errors.js
@@ -553,31 +553,37 @@ function wrapErrorFromDrc(err) {
     if (!err) {
         return err;
     }
+
+    var errMsg = err.message || err.toString();
+
+    // Issue #13 convert all '%' characters, to avoid sprintf formatting errors.
+    errMsg = errMsg.replace(/%/g, '%%');
+
     if (err.name === 'BadDigestError') {
         // Docker registry client digest error.
         return new ValidationFailedError(err,
-            (err.message || err.toString()),
+            errMsg,
             [ {field: 'digest', code: 'Invalid'} ]);
     } else if (err.name === 'ConnectTimeoutError') {
         // Restify connection timeout.
-        return new RemoteSourceError(err, err.message);
+        return new RemoteSourceError(err, errMsg);
     } else if (err.code === 'ECONNREFUSED' || err.code === 'ECONNRESET' ||
         err.code === 'ENOTFOUND')
     {
         // Node.js connection error.
-        return new RemoteSourceError(err, err.message);
+        return new RemoteSourceError(err, errMsg);
     } else if (err.name === 'DownloadError') {
         // Docker registry client download error.
-        return new DownloadError(err, err.message);
+        return new DownloadError(err, errMsg);
     } else if (err.name === 'NotFoundError') {
         // Docker registry client image not found error.
-        return new restify.ResourceNotFoundError(err, err.message);
+        return new restify.ResourceNotFoundError(err, errMsg);
     } else if (err.name === 'UnauthorizedError') {
         // Docker registry client unauthorized error.
-        return new restify.UnauthorizedError(err, err.message);
+        return new restify.UnauthorizedError(err, errMsg);
     } else {
         // Unexpected error - wrap it into an internal error.
-        return new restify.InternalError(err, err.message);
+        return new restify.InternalError(err, errMsg);
     }
 }
 
diff --git a/package.json b/package.json
index 96b024c..24062c7 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "imgapi",
   "description": "Image API to manage images for SDC 7",
-  "version": "4.0.3",
+  "version": "4.0.4",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
-- 
2.21.0

